compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -600,6 +601,25 @@ object Semantic {
         || (meth eq defn.Object_ne)
         || (meth eq defn.Any_isInstanceOf)
 
+      def checkArgsWithParametricity() =
+        val methodType = atPhaseBeforeTransforms { meth.info.stripPoly }
+        var allArgsPromote = true
+        var singleParamSet = immutable.HashSet.from(methodType.paramInfoss.flatten.map(_.repeatedToSingle))
+        val errors = methodType.paramInfoss.flatten.zip(args).flatMap { (info, arg) =>
```

Comment:
There seems to be no need for `immutable.HashSet`. Will something like the following work?

``` Scala
val allParamTypes = methodType.paramInfoss.flatten.map(_.repeatedToSingle)
val errors = allParamTypes.zip(args).flatMap { (info, arg) =>
```

Reply:
The reason I used the HashSet was to improve the runtime complexity of checking that no other parameter type contains the given type.

Also, just constructing an iterator like you suggest won't work since it needs to iterated over for each parameter. They need to be stored in a separate data collection for this reason. The reason I chose an `immutable.HashSet` is because `Type` can be hashed and the `Set` eliminated duplicates implicitly.  I could use a different collection type if you have a better one in mind.

Reply:
> Also, just constructing an iterator like you suggest won't work since it needs to iterated over for each parameter. They need to be stored in a separate data collection for this reason

It seems possible to use `allParamTypes` for the same purpose. Note that method parameters are usually small, and the chance that they have the same type in our setting is even smaller.  Introducing HashSet here seems to be an unnecessary optimization.

Reply:
I just checked the function signatures for `flatten` and `map`. I mistakenly assumed that they did not return a `List`. The approach you are proposing should work!